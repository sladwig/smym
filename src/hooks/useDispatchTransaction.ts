import { useStateRef } from './useStateRef';
import { useState, useCallback } from 'react';
import { EditorState } from 'prosemirror-state';

export function useEditorState(init: EditorState) {
    const [state, setEditorState] = useState<EditorState>(init);
    const stateRef = useStateRef(state);
    const dispatchTransaction = useCallback(
        (tx: any) => setEditorState(stateRef.current.apply(tx)),
        [setEditorState, stateRef],
    );
    return { state, setEditorState, dispatchTransaction, stateRef };
}
