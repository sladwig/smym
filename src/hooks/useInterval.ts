import { useRef, useEffect } from 'react';

// from good ol' Dan, see https://overreacted.io/making-setinterval-declarative-with-react-hooks/
export function useInterval(callback: () => any, delay: number | null) {
    const savedCallback = useRef(() => null);

    useEffect(() => {
        savedCallback.current = callback;
    }, [callback]);

    useEffect(() => {
        const tick = () => savedCallback.current();

        if (delay !== null) {
            const id = setInterval(tick, delay);
            return () => clearInterval(id);
        }
    }, [delay]);
}
