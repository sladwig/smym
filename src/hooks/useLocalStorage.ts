export function useLocalStorage(key: string, fallback: any = null) {
    return localStorage.getItem(key) || fallback;
}

export function useHasInLocalStorage(key: string): boolean {
    return !!localStorage.getItem(key);
}
