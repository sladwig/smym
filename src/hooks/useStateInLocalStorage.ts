import { useState, useEffect } from 'react';

export function useStateInLocalStorage(key: string, defaultValue: any) {
    const [value, setValue] = useState(window.localStorage.getItem(key) || defaultValue);
    useEffect(() => {
        window.localStorage.setItem(key, value);
    }, [value, key]);
    return [value, setValue];
}
