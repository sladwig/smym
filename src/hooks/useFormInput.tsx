import { useState } from 'react';
import { useStateInLocalStorage } from './useStateInLocalStorage';

export function useFormInput(defaultValue: any) {
    const [value, setValue] = useState(defaultValue);

    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setValue(event.target.value);
    };
    const Input = () => <input type="text" value={value} onChange={onChange} />;
    return [value, setValue, Input];
}

export function useLocalStoreBackedFormInput(key: string, defaultValue: any) {
    const [value, setValue] = useStateInLocalStorage(key, defaultValue);

    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setValue(event.target.value);
    };
    const Input = () => <input type="text" value={value} onChange={onChange} />;
    return [value, setValue, Input];
}
