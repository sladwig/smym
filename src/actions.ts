import { emit } from './hooks/useEvents';
import { eventNames } from 'cluster';

export enum Do {
    'tokenInputFocusUpdate' = 'token-input-focus/update',
    'valueUpdate' = 'value/update',
    'positionUpdate' = 'position/update',
}

export type EP = {
    [Do.tokenInputFocusUpdate]: boolean;
    [Do.valueUpdate]: string;
    [Do.positionUpdate]: number;
};

// Emitter
const e = <T extends keyof EP>(eventName: T) => {
    return (value: EP[T]) => emit(eventName, value);
};
// Lazy Emitter
const le = <T extends keyof EP>(eventName: T) => {
    return (value: EP[T]) => () => emit(eventName, value);
};

export const tokenInputHasFocus = le(Do.tokenInputFocusUpdate);
export const tokenInputFocus = tokenInputHasFocus(true);
export const tokenInputBlur = tokenInputHasFocus(false);

export const valueUpdate = e(Do.valueUpdate);
export const positionUpdate = e(Do.positionUpdate);
