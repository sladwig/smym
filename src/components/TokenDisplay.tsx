import React from 'react';
import './TokenDisplay.css';
import { WithDecorations } from './WithDecorations';
import classnames from 'classnames';
import { tokenInputFocus } from '../actions';

interface IProps {}
export const TokenDisplay = ({}: IProps) => {
    return (
        <div onClick={tokenInputFocus} className={classnames('token-input')}>
            <WithDecorations />
        </div>
    );
};
