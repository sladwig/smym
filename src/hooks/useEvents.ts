import NanoEvents from 'nanoevents';
import { useEffect, useMemo } from 'react';
import { EP } from '../actions';

export const emitter = new NanoEvents<EP>();
export const on = emitter.on.bind(emitter);
export const emit = emitter.emit.bind(emitter);

export const useEventEmitter = (): {
    on: typeof emitter.on;
    emit: typeof emitter.emit;
    emitter: typeof emitter;
} => {
    return useMemo(
        () => ({
            on: emitter.on.bind(emitter),
            emit: emitter.emit.bind(emitter),
            emitter,
        }),
        [],
    );
};

export function useOn<EventName extends keyof EP>(
    eventName: EventName,
    cb: (args: EP[EventName]) => any,
) {
    const { on } = useEventEmitter();
    useEffect(() => on(eventName, cb), [eventName, cb, on]);
}

export const useEmit = () => useEventEmitter().emit;
