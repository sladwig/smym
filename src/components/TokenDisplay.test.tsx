import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { TokenDisplay } from './TokenDisplay';
import { emitter } from '../hooks/useEvents';
import unbindAll from 'nanoevents/unbind-all';
import { valueUpdate, EP, Do } from '../actions';

const mockOn = (eventName: keyof EP) => {
    const callbackMock = jest.fn();
    emitter.on(eventName, callbackMock);
    return callbackMock;
};

afterAll(() => unbindAll(emitter));

describe('TokenDisplay', () => {
    test('renders', () => {
        const { container } = render(<TokenDisplay />);
        expect(container).toMatchSnapshot();
    });

    test('emits event input/focus onClick', () => {
        const cb = mockOn(Do.tokenInputFocusUpdate);
        const { container } = render(<TokenDisplay />);

        fireEvent.click(container.firstChild as HTMLInputElement);
        expect(cb).toHaveBeenCalledTimes(1);
    });

    test('reacts on value/update event', () => {
        const { container } = render(<TokenDisplay />);

        valueUpdate('oh');
        expect((container.firstChild as HTMLDivElement).innerHTML).toMatchSnapshot();
    });

    test('should not render anything when value is empty', () => {
        const { container } = render(<TokenDisplay />);
        expect((container.firstChild as HTMLDivElement).innerHTML).toBeEmpty();
    });

    test('should render caret on focus when value is empty', () => {
        const { container } = render(<TokenDisplay />);
        expect((container.firstChild as HTMLDivElement).innerHTML).toMatchSnapshot();
    });
});
