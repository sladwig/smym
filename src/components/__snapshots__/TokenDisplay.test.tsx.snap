// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenDisplay reacts on value/update event 1`] = `"<span><span class=\\"action-decorator\\"><span class=\\"click-position-emitter\\"><span class=\\"position-emitter\\">o<span class=\\"position-area-overlay\\"><span class=\\"position-area\\"></span><span class=\\"position-area\\"></span></span></span></span><span class=\\"click-position-emitter\\"><span class=\\"position-emitter\\">h<span class=\\"position-area-overlay\\"><span class=\\"position-area\\"></span><span class=\\"position-area\\"></span></span></span></span><svg>cross-icon.svg</svg></span></span>"`;

exports[`TokenDisplay renders 1`] = `
<div>
  <div
    class="token-input"
  />
</div>
`;

exports[`TokenDisplay should render caret on focus when value is empty 1`] = `""`;
